### ISSUE TITLE
History is bloating up state

### ISSUE THREAD

**Issue by michelengelen** [omnidan/redux-undo#211]
First of all I would like to thank the contributors for this awesome library. 
It really helped me a lot in designing my application.

But (there is always a BUT) as far as I can see there is one performance issue that shouldn't be underestimated.

With consecutive calls to the included reducers the state gets really bloated up, because with every action a new level of `history` gets added making it really big if you add up several hundred actions.

In my opinion I don't really see the need for this specific history property, since every past and future state is getting pushed/sliced from the respective `past` and `future` properties anyways. So why is there this history in the first place? If I am mistaken for some reason feel free to lecture me about it.

<img width="722" alt="screen shot 2018-08-08 at 10 05 42" src="https://user-images.githubusercontent.com/32863416/43824570-a686fa42-9af2-11e8-854c-9d85e1b06aa2.png">

This object has several hundred levels of encapsulated history states. This will potentially slow down my application significantly.

I would gladly offer my help to improve on this issue.

**Comment by craPkit** [comment]
Here's a microscopic reducer to work around said bloat:

```
function patchHistoryBloat(reducer) {
  return (state, action, ...rest) => {
    const res = reducer(state, action, ...rest);
    if (res.history.history) {
      delete res.history.history;
    }
    return res;
  };
}

const myReducer = patchHistoryBloat(undoable(myOriginalReducer))
```

**Comment by yairEO** [comment]
What you actually want is to only save the *diff* between states, so the *current state* always represents a  "full" state and every past or future state will automatically be saved only as a diff. 

While this guarantees beautiful-looking state history Object, I am not sure if it's better this way in regarding the performance of having to deep-diff the state in opposed to simply bloating up the history state object with full copies of the state as-it-was at that moment. I think it really depends of the kind of data being saved (length & depth)

**Comment by lidaof** [comment]
I guess you are using 0.6? I think 1.0 beta seems solved this, but I always get cannot find module error.

**Comment by omnidan** [comment]
@michelengelen 1.0-beta should solve the issue of nested histories. however, there is still the issue of storing the full history instead of a diff between states (as @yairEO has pointed out)

@lidaof 1.0-beta uses the new import system, which means that if you use `require()` you need to do `require('redux-undo').default`: https://github.com/omnidan/redux-undo#note-on-100-beta7

**Comment by nmay231** [comment]
Since the bug in the original question is fixed in 1.0-beta, this issue will be closed.

As for storing history as a list of diffs (instead of the full history), there is already an issue (#36) as well as some discussion in #145.
